File: go-version/leetcode/LC2141/max_run_time.go
package LC2141

/**
 * @Author: cutejiuge cutejiuge@163.com
 * @Date: 2025/12/1 上午9:12
 * @FilePath: max_run_time
 * @Description:
 */

func maxRunTime(n int, batteries []int) int64 {
	var total int64
	for _, num := range batteries {
		total += int64(num)
	}

	low, high := int64(0), total/int64(n)+1
	for low+1 < high {
		mid := low + (high-low)/2
		var temp int64
		for _, num := range batteries {
			temp += min(mid, int64(num))
		}
		if int64(n)*mid <= temp {
			low = mid
		} else {
			high = mid
		}
	}
	return low
}

File: python-version/LC2141/max_run_time.py
from typing import List


class Solution:
    @staticmethod
    def maxRunTime(n: int, batteries: List[int]) -> int:
        total = sum(batteries)
        low, high = 0, total // n + 1
        while low + 1 < high:
            mid = low + (high - low) // 2
            temp = 0
            for num in batteries:
                temp += min(mid, num)
            if n * mid <= temp:
                low = mid
            else:
                high = mid
        return low
